### PR TITLE
Add quantization (dynamic)

### PR DESCRIPTION
**Pull request by avishreekh**
Solves #72

### CODE REVIEW

**Comment by Het-Shah** [comment]
So we need to check whether this is working or not right? So benchmark it this way train a ResNet normally on MNIST or any dataset report its accuracy. Then do quantization and then report accuracy after quantization. Also, report the size of weights before and after quantization.

**Comment by avishreekh** [comment]
> So we need to check whether this is working or not right? So benchmark it this way train a ResNet normally on MNIST or any dataset report its accuracy. Then do quantization and then report accuracy after quantization. Also, report the size of weights before and after quantization.

Sure

File: logs.rst [hunk 1]
```diff
@@ -69,3 +69,15 @@ Results for Pruning Algorithms -
...
+Results for  Dynamic Quantization - 
+===================================
+
++------------+---------+---------------+----------------+-----------------+----------+--------------+
+| Model Used | Dataset |    Model      |    Size (MB)   | Training epochs | Accuracy | Time elapsed |
++============+=========+===============+================+=================+==========+==============+
+|            |         |     Base      |       4.15     |                 |  0.9582  |  28.84 s     |
+|            |         +---------------+----------------+                 +----------+--------------+
+|  ResNet152 |  MNIST  |   Quantized   |       0.56     |        1        |  0.9583  |  28.71 s     |
++------------+---------+---------------+----------------+-----------------+----------+--------------+
```
Comment (Het-Shah) [inline logs.rst:82]: Can you explain the columns and the whole process that you followed once?

Comment (Het-Shah) [reply logs.rst:82]: I don't understand the time elapsed column for example

Comment (avishreekh) [reply logs.rst:82]: Oh sure.
Process: I trained a ResNet152 for one epoch on MNIST. After training it was quantized. The accuracy obtained by the non-quantized model was 0.9582 and it took 28.84 secs for it to run on the test set on CPU. The quantized model achieved an accuracy of 0.9583 and it took 28.71 secs for it to run on the test set on CPU.
Time elapsed is the time the model took to run on the test set.

Here is a link to a similar analysis by PyTorch:
https://pytorch.org/tutorials/advanced/dynamic_quantization_tutorial.html

Comment (Het-Shah) [reply logs.rst:82]: How many epochs did you train ResNet for?

Comment (avishreekh) [reply logs.rst:82]: One epoch

Comment (Het-Shah) [reply logs.rst:82]: Can you perform a similar analysis for CIFAR-10 with more training epochs maybe?

Comment (avishreekh) [reply logs.rst:82]: I wouldn't mind but why is that needed?

**Comment by Het-Shah** [comment]
Now that we have figured out that this is working. We should probably add it to the Base KD class as well using `*args`.

**Comment by avishreekh** [comment]
And what about the training on CIFAR?

**Comment by Het-Shah** [comment]
> And what about the training on CIFAR?

Yes this too obviously. 

Additionally, we should add some helper functions to report results for quantization. For example, the model size calculation and inference time of non-quantized and quantized model.

**Comment by avishreekh** [comment]
> > And what about the training on CIFAR?
> 
> Yes this too obviously.
> 
> Additionally, we should add some helper functions to report results for quantization. For example, the model size calculation and inference time of non-quantized and quantized model.

Should I add these to quantization class?

**Comment by Het-Shah** [comment]
> Should I add these to quantization class?

Yes. Do you think anything else is needed?

**Comment by avishreekh** [comment]
No I believe that should work

**Comment by Het-Shah** [comment]
Hmm ok then

**Comment by avishreekh** [comment]
> And what about the training on CIFAR?

Can I do this on pretrained ResNet from torchvision models?

**Comment by avishreekh** [comment]
Check this and let me know if it looks good

File: KD_Lib/Quantization/dynamic/dynamic_quantization.py [hunk 2]
```diff
@@ -9,14 +10,35 @@ class Dynamic_Quantizer:
...
+        print(f"Size of quantized_model (MB): {quantized_size}")
+
+
+### Helper Functions
+
+
+def get_size_of_model(model):
+    torch.save(model.state_dict(), "temp.p")
+    model_size = os.path.getsize("temp.p") / 1e6
+    os.remove("temp.p")
```
Comment (Het-Shah) [inline KD_Lib/Quantization/dynamic/dynamic_quantization.py:44]: Move this to the class itself with `_get_size_of_model()` name coz its a helper function

Comment (avishreekh) [reply KD_Lib/Quantization/dynamic/dynamic_quantization.py:44]: But that function is not essential to that class

**Review by Het-Shah — COMMENTED** [review]
This seems good now.

**Comment by avishreekh** [comment]
Ok so too many changes in a PR. Sorry for that.
Major features:
1. Added a quantization base class with functionalities for comparing sizes and inference performance.
2. Added static quantization module

File: KD_Lib/Quantization/common/base_class.py [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+        self.quantized_model = model
+
+    def quantize(self):
+        """
+		Function used for quantization
+		"""
+        raise NotImplementedError
+
+    def compare_model_sizes(self):
+        """
```
Comment (Het-Shah) [inline KD_Lib/Quantization/common/base_class.py:23]: a better name would be `get_model_sizes()` from users perspective

File: KD_Lib/Quantization/common/base_class.py [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+        original_size = self._get_size_of_model(self.model)
+        quantized_size = self._get_size_of_model(self.quantized_model)
+
+        print("-" * 80)
+        print(f"Size of original model (MB): {original_size}")
+        print(f"Size of quantized_model (MB): {quantized_size}")
+
+    def compare_inference_performance(self, data_loader):
+        """
```
Comment (Het-Shah) [inline KD_Lib/Quantization/common/base_class.py:34]: Same here

File: KD_Lib/Quantization/static/static_quantization.py [hunk 5]
```diff
@@ -0,0 +1,66 @@
...
+        num_calibration_batches=10,
+        qconfig=torch.quantization.default_qconfig,
+    ):
+        """
+		Function used for quantization
+        :param data_loader(torch.utils.data.DataLoader): DataLoader used for calibration
+        :param criterion(torch Loss_fn): Loss function used for calibration
+        :param num_calibration_batches(int): Number of batches used for calibration
+        :param qconfig: Configuration used for quantization
+		"""
```
Comment (Het-Shah) [inline KD_Lib/Quantization/static/static_quantization.py:28]: Add an empty line between description and params, otherwise autodoc won't pick it up properly

File: KD_Lib/Quantization/static/static_quantization.py [hunk 6]
```diff
@@ -0,0 +1,66 @@
...
+        self._calibrate_model(data_loader, num_calibration_batches, criterion)
+
+        print("Converting to quantized model...")
+        torch.quantization.convert(self.quantized_model, inplace=True)
+
+        return self.quantized_model
+
+    def _calibrate_model(self, data_loader, num_batches, criterion):
+        self.quantized_model.eval()
+        correct = 0
```
Comment (Het-Shah) [inline KD_Lib/Quantization/static/static_quantization.py:46]: Add docstrings to helper functions as well or else they won't get picked up by autodoc

**Review by Het-Shah — COMMENTED** [review]
Seems like great work to me. Seems ready just small stuff here and there.

File: logs.rst [hunk 7]
```diff
@@ -69,3 +69,19 @@ Results for Pruning Algorithms -
...
+| Model Used | Dataset |    Model      |    Size (MB)   | Training epochs | Accuracy | Time elapsed |
++============+=========+===============+================+=================+==========+==============+
+|  ResNet152 |  MNIST  |     Base      |       4.15     |        1        |  0.9582  |  28.84 s     |
+|            |         +---------------+----------------+                 +----------+--------------+
+|            |         |   Quantized   |       0.56     |                 |  0.9583  |  28.71 s     |
++            +---------+---------------+----------------+-----------------+----------+--------------+
+|            | CIFAR10 |     Base      |       4.15     |        1        |  0.9582  |  28.84 s     |
+|            |         +---------------+----------------+                 +----------+--------------+
+|            |         |   Quantized   |       0.56     |                 |  0.9583  |  28.71 s     |
++------------+---------+---------------+----------------+-----------------+----------+--------------+
```
Comment (Het-Shah) [inline logs.rst:86]: One more thing here it will be good to mention which method did we use for quantization.

Comment (Het-Shah) [reply logs.rst:86]: How are both MNIST and CIFAR10 getting the same metrics :P

File: KD_Lib/Quantization/common/base_class.py (hunk 3 above)
Comment (avishreekh) [reply KD_Lib/Quantization/common/base_class.py:23]: I thought of the same previously. However, named it this way coz it will be used for comparison. I'll change it though, `get_model_sizes` seems nicer

File: logs.rst (hunk 7 above)
Comment (avishreekh) [reply logs.rst]: They are not lol. I just copy pasted to enter the result later but forgot

**Comment by avishreekh** [comment]
Why is this build failing?

**Comment by Het-Shah** [comment]
black

**Comment by avishreekh** [comment]
I have done that

**Review by Het-Shah — APPROVED** [review]
